### PR TITLE
Add fuzzy-date function to case search

### PR DESCRIPTION
**Pull request by MartinRiese**
## Product Description

Add a function to fuzzy search dates.

## Technical Summary

https://dimagi.atlassian.net/browse/USH-4861

## Feature Flag

Case search needs to be enabled

### Automated test coverage

Added new unit test

### QA Plan

Update Potential Duplicate Clients case list in staging and perform search.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Comment by MartinRiese** [comment]
@esoergel No tests or documentation yet. But I was curious about your thoughts on this? I was think about adding a third parameter to let you specify a range in years for how far the permutations could be away from the original date.

File: corehq/apps/case_search/xpath_functions/query_functions.py [hunk 1]
```diff
@@ -95,6 +97,50 @@ def fuzzy_match(node, context):
...
+    confirm_args_count(node, 2)
+    property_name = _property_name_to_string(node.args[0], node)
+    value = unwrap_value(node.args[1], context)
+
+    return case_property_query(property_name, date_permutations(value))
+
+
+def date_permutations(date_str):
+    [year, month, day] = date_str.split('-')
+    decade_flipped = year[:2:] + year[2::][::-1]
```
Comment (esoergel) [inline corehq/apps/case_search/xpath_functions/query_functions.py:111]:
```suggestion
    decade_flipped = year[:2] + year[:1:-1]
```
or 
```suggestion
    decade_flipped = year[:2] + year[3] + year[2]
```

File: corehq/apps/case_search/xpath_functions/query_functions.py [hunk 2]
```diff
@@ -95,6 +97,50 @@ def fuzzy_match(node, context):
...
+        f"{year}-{reverse_day}-{reverse_month}",
+        f"{decade_flipped}-{month}-{day}",
+        f"{decade_flipped}-{day}-{month}",
+        f"{decade_flipped}-{reverse_month}-{day}",
+        f"{decade_flipped}-{day}-{reverse_month}",
+        f"{decade_flipped}-{month}-{reverse_day}",
+        f"{decade_flipped}-{reverse_day}-{month}",
+        f"{decade_flipped}-{reverse_month}-{reverse_day}",
+        f"{decade_flipped}-{reverse_day}-{reverse_month}"
+    ]
```
Comment (esoergel) [inline corehq/apps/case_search/xpath_functions/query_functions.py:131]: This list is super readable and makes this function's implementation clear - nice work!

**Review by esoergel — COMMENTED** [review]
Implementation looks great, and the profiling results so far sound quite promising

File: corehq/apps/case_search/xpath_functions/query_functions.py [hunk 3]
```diff
@@ -95,6 +97,50 @@ def fuzzy_match(node, context):
     return case_property_query(property_name, value, fuzzy=True)
 
 
+def fuzzy_date(node, context):
+    """fuzzy-match(dob, '2024-12-03')"""
+    confirm_args_count(node, 2)
+    property_name = _property_name_to_string(node.args[0], node)
+    value = unwrap_value(node.args[1], context)
+
+    return case_property_query(property_name, date_permutations(value))
```
[the remark this answers is not in the document]
Comment (MartinRiese) [reply corehq/apps/case_search/xpath_functions/query_functions.py:106]: Just to clarify: you mean I have two cases. One with 2024-12-03 and one with 2024-03-12. If I search for 2024-12-03 I should guaranteed that the first one shows up first/ is ranked higher? Are the case search results sorted by the ES score?

File: corehq/apps/case_search/xpath_functions/query_functions.py (hunk 1 above)
Comment (MartinRiese) [reply corehq/apps/case_search/xpath_functions/query_functions.py:111]: Agreed. Less colons is better :)

File: corehq/apps/case_search/xpath_functions/query_functions.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/apps/case_search/xpath_functions/query_functions.py:106]: Yeah I believe that's the case.  Good to confirm with testing though.

File: corehq/apps/es/case_search.py [hunk 4]
```diff
@@ -233,7 +233,7 @@ def multiplex_to_adapter(domain):
     return None
 
 
-def case_property_query(case_property_name, value, fuzzy=False, multivalue_mode=None):
+def case_property_query(case_property_name, value, fuzzy=False, multivalue_mode=None, boost_first=False):
```
Comment (MartinRiese) [inline corehq/apps/es/case_search.py:236]: @esoergel I am feeling a bit iffy about adding yet another flag here. I see two alternatives:
1) just create a separate function `case_property_query_boost_first` and call it from `fuzzy_date`
2) use `_base_property_query` directly in `fuzzy_date` probably renaming it without the underscore in the process.
I am curious, if you have a preference?

File: corehq/apps/case_search/tests/test_filter_dsl.py [hunk 5]
```diff
@@ -206,6 +206,21 @@ def test_fuzzy_match(self):
...
+                "2024-12-03",
+                "2024-03-12",
+                "2024-03-21",
+                "2024-12-30",
+                "2042-12-03",
+                "2042-03-12",
+                "2042-03-21",
+                "2042-12-30"
+            ], boost_first=True)
+        )
```
[the remark this answers is not in the document]
Comment (MartinRiese) [reply corehq/apps/case_search/tests/test_filter_dsl.py:222]: @AddisonDunn is there a test the run ES queries against test data already? I am having trouble finding any.

**Review by AddisonDunn — APPROVED** [review]

File: corehq/apps/case_search/tests/test_filter_dsl.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (MartinRiese) [reply corehq/apps/case_search/tests/test_filter_dsl.py:222]: @AddisonDunn pushed the test

**Review by AddisonDunn — APPROVED** [review]

File: corehq/apps/es/tests/test_case_search_es.py [hunk 6]
```diff
@@ -435,6 +436,21 @@ def test_fuzzy_case_property_query(self):
             ['c3']
         )
 
+    def test_fuzzy_date(self):
+        print("############## test_fuzzy_date")
```
[the remark this answers is not in the document]
Comment (MartinRiese) [reply corehq/apps/es/tests/test_case_search_es.py]: I do not.

**Review by AddisonDunn — APPROVED** [review]

**Pull request merged**